src/blocks/store/adl.clj

```diff
@@ -196,18 +196,17 @@
 
   (-put!
     [this block]
-    (or (.-get this (:id block))
-        (let [path (id->path root (:id block))]
-          (with-open [output (.createFile client path IfExists/FAIL write-permission true)
-                      content (block/open block)]
-            (io/copy content output))
-          (try-until #(= (:size block)
-                         (or (.length (.getDirectoryEntry client path)) 0))
-                     {:tries 5 :wait-period 200 :intent "upload complete"})
-          (.setPermission client path read-permission)
-          (try-until #(= read-permission
-                         (.-octalPermissions (.getAclStatus client path)))
-                     {:intent "permission flag set"}))))
+    (let [path (id->path root (:id block))
+          pre-write-path (str path pre-write-path-suffix)]
+      (when-not (.checkExists client path)
+        (with-open [output (.createFile client pre-write-path IfExists/FAIL write-permission true)
+                    content (block/open block)]
+          (io/copy content output))
+        (try-until #(= (:size block)
+                       (or (.length (.getDirectoryEntry client pre-write-path)) 0))
+                   {:tries 5 :wait-period 200 :intent "upload complete"})
```

Comment:
The rename approach is fine by me if it's more atomic. I think there's still value in setting the file to be read-only after upload, given that this is content-addressed data.